pkg/minishift/docker/image/oci_image_handler.go

```diff
@@ -331,21 +334,54 @@ func (handler *OciImageHandler) exportImage(image string, config *ImageCacheConf
 		return fmt.Errorf("Invalid image source '%s': %v", image, err)
 	}
 
-	destRef, err := layout.NewReference(config.HostCacheDir, image)
+	// tmpImageIndexLocation should be a directory location which will be atomic for each image.
+	// for an image "openshift/origin-control-plane:v3.10.0"
+	// it will be $HOME/.minishift/cache/image/openshift-origin-control-plane-v3.10.0
+	r := strings.NewReplacer(":", "-", "/", "-")
+	tmpImageIndexLocation := filepath.Join(config.HostCacheDir, r.Replace(image))
+
+	destRef, err := layout.NewReference(tmpImageIndexLocation, image)
 	if err != nil {
 		return fmt.Errorf("Invalid image destination '%v': %v", destRef, err)
 	}
 
-	err = handler.copyImage(srcRef, destRef, policyContext)
+	err = handler.copyImage(srcRef, destRef, policyContext, config.HostCacheDir)
+	if err != nil {
+		return err
+	}
+
+	// Get index of current export image
+	pulledImageIndex, err := handler.getIndex(tmpImageIndexLocation)
 	if err != nil {
 		return err
 	}
 
+	// Remove tmpImageIndexLocation
+	if err := os.RemoveAll(tmpImageIndexLocation); err != nil {
+		return err
+	}
+
+	// Get index of already available image
+	availableImageIndex, err := handler.getIndex(config.HostCacheDir)
+	if err != nil {
+		return err
+	}
+
+	if availableImageIndex == nil {
+		availableImageIndex = &Index{Manifests: Manifests{}, SchemaVersion: 2}
+	}
+
+	availableImageIndex.Manifests = append(availableImageIndex.Manifests, pulledImageIndex.Manifests...)
```

Comment:
Is this atomic operation? I do not think so. If two process are exactly at this step at exact same time how will it behave? However I agree the whole PR  reduces the probability of overwriting the index file. But if you want to go one step ahead you should use mutex to make sure only one process can call this step. 

https://stackoverflow.com/questions/46138395/how-make-a-function-thread-safe-in-golang

Reply:
@LalatenduMohanty what is the probability you think this has to behave as unexpected, check the flow. 

- the image first goes to index on its own.
- Take the existing index in memory and append the index.
- In meanwhile if some other process is running then it will also take index in the memory and append it. We are not overwriting anything at all so I don't know why we want to use a mutex or any other complicated stuff.